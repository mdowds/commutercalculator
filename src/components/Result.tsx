import * as React from "react";
import { JourneyResult, SeasonTickets, Travelcard } from "../types";
import { CSSProperties } from "react";

interface ResultProps {
  readonly result: JourneyResult;
  readonly isSelected: boolean;
  readonly averageHousePrice?: number;
  onSelectResult(selectedResult: JourneyResult): void;
}

export default class Result extends React.Component<ResultProps, {}> {
  travelcardDescription(travelcard: Travelcard) {
    return (
      <div>
        Zone {travelcard.minZone} to {travelcard.maxZone} travelcard: £
        {travelcard.price}
      </div>
    );
  }

  housePriceDescription() {
    const outcode = this.props.result.origin.postcode.split(" ")[0];
    const price = this.props.averageHousePrice
      ? "£" + this.props.averageHousePrice.toLocaleString("en-GB")
      : "Loading";
    return (
      <div>
        Average house price for {outcode}: {price}
      </div>
    );
  }

  makeDetails(result: JourneyResult) {
    const travelcardDiv = result.seasonTickets.travelcard
      ? this.travelcardDescription(result.seasonTickets.travelcard)
      : null;
    const seasonTicketDiv = result.seasonTickets.seasonTicket ? (
      <div>
        Direct season ticket: £{result.seasonTickets.seasonTicket.price}
      </div>
    ) : null;

    const seasonTicketsDescription = this.seasonTicketsHasValues(
      result.seasonTickets
    ) ? (
      <div>
        {travelcardDiv}
        {seasonTicketDiv}
      </div>
    ) : (
      <div>Season ticket not found</div>
    );

    const directionsLink = (
      <span>
        <a href={result.directionsUrl}>View directions on Google Maps</a> (May
        show a different journey time)
      </span>
    );

    return (
      <div>
        {seasonTicketsDescription}
        {this.housePriceDescription()}
        {directionsLink}
      </div>
    );
  }

  seasonTicketsHasValues(seasonTickets: SeasonTickets) {
    return seasonTickets.travelcard || seasonTickets.seasonTicket;
  }

  render() {
    const bgColour = this.props.isSelected ? "#ddd" : "white";

    const divStyle: CSSProperties = {
      padding: 10,
      borderBottom: "1px solid grey",
      backgroundColor: bgColour
    };

    const details = this.props.isSelected ? (
      <div style={{ margin: "10px 0px 0px 30px", fontSize: "75%" }}>
        {this.makeDetails(this.props.result)}
      </div>
    ) : null;

    const icon = this.props.isSelected
      ? "img/chevron-up.svg"
      : "img/chevron-down.svg";

    return (
      <div
        style={divStyle}
        onClick={() => {
          this.props.onSelectResult(this.props.result);
        }}
      >
        <div>
          <div style={{ float: "left", marginRight: 10 }}>
            <img
              alt="Expand results"
              src={icon}
              style={{ width: 20, height: 20 }}
            />
          </div>
          <div style={{ display: "inline-block", width: "65%" }}>
            {this.props.result.origin.name}
          </div>
          <div style={{ float: "right" }}>
            {this.props.result.journeyTime} mins
          </div>
        </div>
        {details}
      </div>
    );
  }
}
