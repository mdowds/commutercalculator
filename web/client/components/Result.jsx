"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
function travelcardDescription(travelcard) {
    return <div>Zone {travelcard.minZone} to {travelcard.maxZone} travelcard: £{travelcard.price}</div>;
}
function makeDetails(result) {
    var travelcardDiv = result.seasonTickets.travelcard ? travelcardDescription(result.seasonTickets.travelcard) : null;
    var seasonTicketDiv = result.seasonTickets.seasonTicket ? <div>Direct season ticket: £{result.seasonTickets.seasonTicket.price}</div> : null;
    var seasonTicketsDescription = <div>
        {travelcardDiv}
        {seasonTicketDiv}
    </div>;
    return seasonTicketsHasValues(result.seasonTickets) ?
        seasonTicketsDescription :
        <div>Season ticket not found</div>;
}
function seasonTicketsHasValues(seasonTickets) {
    return seasonTickets.travelcard || seasonTickets.seasonTicket;
}
function Result(props) {
    var bgColour = props.isSelected ? '#ddd' : 'white';
    var divStyle = {
        padding: 10,
        borderBottom: "1px solid grey",
        backgroundColor: bgColour
    };
    var details = props.isSelected ? <div style={{ margin: '10px 0px 0px 30px', fontSize: '75%' }}>{makeDetails(props.result)}</div> : null;
    var icon = props.isSelected ? "img/chevron-up.svg" : "img/chevron-down.svg";
    return (<div style={divStyle} onClick={function () { props.onSelectResult(props.result); }}>
            <div>
                <div style={{ float: "left", marginRight: 10 }}><img src={icon} style={{ width: 20, height: 20 }}/></div>
                <div style={{ display: "inline-block", width: "65%" }}>{props.result.origin.name}</div>
                <div style={{ float: "right" }}>{props.result.journeyTime} mins</div>
            </div>
            {details}
        </div>);
}
exports.default = Result;
