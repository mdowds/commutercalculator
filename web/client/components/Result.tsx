import * as React from 'react';
import {JourneyResult, Station} from '../types';
import {CSSProperties} from "react";

interface ResultProps {
    readonly result: JourneyResult,
    readonly isSelected: boolean,
    onSelectResult(selectedResult: string): void
}

function makeDetails(result: JourneyResult) {
    return result.seasonTicket ? <div>Season ticket price: Â£{result.seasonTicket.price}</div> : null;
}

export default function Result(props: ResultProps) {

    const bgColour = props.isSelected ? '#ddd' : 'white';

    const divStyle: CSSProperties = {
        padding: 10,
        borderBottom: "1px solid grey",
        backgroundColor: bgColour
    };

    const details = props.isSelected ? <div style={{margin: '10px 0px 0px 30px', fontSize: '75%'}}>{makeDetails(props.result)}</div> : null;

    const icon = props.isSelected ? "img/chevron-up.svg" : "img/chevron-down.svg";

    return (
        <div style={divStyle} onClick={() => {props.onSelectResult(props.result.origin.id)}}>
            <div>
                <div style={{float: "left", marginRight: 10}}><img src={icon} style={{width: 20, height: 20}}/></div>
                <div style={{display: "inline-block", width: "65%"}}>{props.result.origin.name}</div>
                <div style={{float: "right"}}>{props.result.journeyTime} mins</div>
            </div>
            {details}
        </div>
    );
}
