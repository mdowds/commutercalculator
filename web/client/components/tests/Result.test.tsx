import * as React from 'react';
import * as renderer from 'react-test-renderer';
import {} from 'jest';
import Result from '../Result';

const mockResult = {origin: {id: "FOO", name: "Foo", position:{lat: 0, lng: 0}}, journeyTime: 10, seasonTicket: {price: 1000}};

test("Result renders the result", () => {
    const result = renderer.create(<Result result={mockResult} isSelected={false} onSelectResult={() => {}} />);
    expect(result.toJSON()).toMatchSnapshot();
});

test("Result renders the result with details", () => {
    const result = renderer.create(<Result result={mockResult} isSelected={true} onSelectResult={() => {}} />);
    expect(result.toJSON()).toMatchSnapshot();
});
