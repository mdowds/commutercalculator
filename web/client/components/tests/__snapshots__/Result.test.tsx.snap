// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Result renders the result 1`] = `
<div
  onClick={[Function]}
  style={
    Object {
      "backgroundColor": "white",
      "borderBottom": "1px solid grey",
      "padding": 10,
    }
  }
>
  <div>
    <div
      style={
        Object {
          "float": "left",
          "marginRight": 10,
        }
      }
    >
      <img
        src="img/chevron-down.svg"
        style={
          Object {
            "height": 20,
            "width": 20,
          }
        }
      />
    </div>
    <div
      style={
        Object {
          "display": "inline-block",
          "width": "65%",
        }
      }
    >
      Foo
    </div>
    <div
      style={
        Object {
          "float": "right",
        }
      }
    >
      10
       mins
    </div>
  </div>
</div>
`;

exports[`Result renders the result with details 1`] = `
<div
  onClick={[Function]}
  style={
    Object {
      "backgroundColor": "#ddd",
      "borderBottom": "1px solid grey",
      "padding": 10,
    }
  }
>
  <div>
    <div
      style={
        Object {
          "float": "left",
          "marginRight": 10,
        }
      }
    >
      <img
        src="img/chevron-up.svg"
        style={
          Object {
            "height": 20,
            "width": 20,
          }
        }
      />
    </div>
    <div
      style={
        Object {
          "display": "inline-block",
          "width": "65%",
        }
      }
    >
      Foo
    </div>
    <div
      style={
        Object {
          "float": "right",
        }
      }
    >
      10
       mins
    </div>
  </div>
  <div
    style={
      Object {
        "fontSize": "75%",
        "margin": "10px 0px 0px 30px",
      }
    }
  >
    <div>
      Season ticket price: Â£
      1000
    </div>
  </div>
</div>
`;
